Fix failing unit tests, fail on deprecation warning qiskit/main

We run our unit [tests](https://github.com/Qiskit/qiskit-ibm-runtime/actions/workflows/unit-tests-terra-main.yml) on a nightly cron job again the main branch of qiskit (terra)

- [x] check/fix all failing tests 
- [x] check/fix all deprecation warnings - we should also fail the test on deprecation warning 


cc @1ucian0 @drew-distefano

Reopening because 2 tests are still failing 

https://github.com/Qiskit/qiskit-ibm-runtime/actions/runs/6263642457/job/17008487156#step:5:839

`test_coder_qc`
```
NotImplementedError: This gate class <class 'qiskit.circuit.library.standard_gates.h.HGate'> does not support manually setting a condition on an instance. Instead you must set the label when instantiating a new object or via the .c_if() method
```

`test_unsupported_values_for_sampler_options`

```
AssertionError: 'resilience_level' not found in 'The 0-th circuit does not have Measure instruction. Without measurements, the circuit cannot be sampled from.'
```

--- a/test/ibm_test_case.py
+++ b/test/ibm_test_case.py
@@ -17,6 +17,7 @@ import copy
 import logging
 import inspect
 import unittest
+import warnings
 from contextlib import suppress
 from collections import defaultdict
 from typing import DefaultDict, Dict
@@ -45,6 +46,8 @@ class IBMTestCase(unittest.TestCase):
         filename = "%s.log" % os.path.splitext(inspect.getfile(cls))[0]
         setup_test_logging(cls.log, filename)
         cls._set_logging_level(logging.getLogger(QISKIT_IBM_RUNTIME_LOGGER_NAME))
+        # fail test on deprecation warnings from qiskit
+        warnings.filterwarnings("error", category=DeprecationWarning, module="qiskit")
 
     @classmethod
     def _set_logging_level(cls, logger: logging.Logger) -> None:

--- a/test/unit/test_data_serialization.py
+++ b/test/unit/test_data_serialization.py
@@ -96,15 +96,10 @@ class TestDataSerialization(IBMTestCase):
 
         # filter warnings triggered by opflow imports
         with warnings.catch_warnings():
-            warnings.filterwarnings(
-                "ignore", category=DeprecationWarning, module=r"qiskit\.opflow\."
-            )
+            warnings.filterwarnings("ignore", category=DeprecationWarning)
             from qiskit.opflow import PauliSumOp  # pylint: disable=import-outside-toplevel
 
-        # catch warnings triggered by opflow use
-        with warnings.catch_warnings(record=True) as w_log:
             deprecated_op = PauliSumOp(SparsePauliOp(Pauli("XYZX"), coeffs=[2]))
-            self.assertTrue(len(w_log) > 0)
 
         coeff_x = Parameter("x")
         coeff_y = coeff_x + 1
@@ -124,14 +119,7 @@ class TestDataSerialization(IBMTestCase):
                 with warnings.catch_warnings():
                     # filter warnings triggered by opflow imports
                     # in L146 of utils/json.py
-                    warnings.filterwarnings(
-                        "ignore", category=DeprecationWarning, module=r"qiskit\.opflow\."
-                    )
-                    warnings.filterwarnings(
-                        "ignore",
-                        category=DeprecationWarning,
-                        module=r"qiskit_ibm_runtime\.utils\.json",
-                    )
+                    warnings.filterwarnings("ignore", category=DeprecationWarning)
                     decoded = json.loads(encoded, cls=RuntimeDecoder)
                     self.assertEqual(operator, decoded)
 
@@ -141,7 +129,12 @@ class TestDataSerialization(IBMTestCase):
         self.assertIsInstance(noise_model, NoiseModel)
         encoded = json.dumps(noise_model, cls=RuntimeEncoder)
         self.assertIsInstance(encoded, str)
-        decoded = json.loads(encoded, cls=RuntimeDecoder)
+        with warnings.catch_warnings():
+            warnings.filterwarnings(
+                "ignore",
+                category=DeprecationWarning,
+            )
+            decoded = json.loads(encoded, cls=RuntimeDecoder)
         self.assertIsInstance(decoded, NoiseModel)
         self.assertEqual(noise_model.noise_qubits, decoded.noise_qubits)
         self.assertEqual(noise_model.noise_instructions, decoded.noise_instructions)

